File: Data Structure/E3/E3.py
class NoLista:
    def __init__(self, info):
        self.info = info
        self.next = None

    def set_info(self, info):
        self.info = info

    def get_info(self):
        return self.info

    def set_next(self, next):
        self.next = next

    def get_next(self):
        return self.next


class Lista:
    def __init__(self):
        self.first = None

    def insert(self, v):
        new_node = NoLista(v)
        if self.is_empty():
            self.first = new_node
        else:
            current = self.first
            while current.get_next() is not None:
                current = current.get_next()
            current.set_next(new_node)

    def print_list(self):
        current = self.first
        while current is not None:
            print(current.get_info(), end=' ')
            current = current.get_next()
        print()

    def is_empty(self):
        return self.first is None

    def search(self, v):
        current = self.first
        while current is not None:
            if current.get_info() == v:
                return current
            current = current.get_next()
        return None

    def length(self):
        count = 0
        current = self.first
        while current is not None:
            count += 1
            current = current.get_next()
        return count

    def last(self):
        current = self.first
        if current is None:
            return None
        while current.get_next() is not None:
            current = current.get_next()
        return current

    def remove(self, v):
        if self.is_empty():
            return
        if self.first.get_info() == v:
            self.first = self.first.get_next()
            return
        previous = self.first
        current = self.first.get_next()
        while current is not None:
            if current.get_info() == v:
                previous.set_next(current.get_next())
                return
            previous = current
            current = current.get_next()

    def clear(self):
        self.first = None

    def insert_end(self, v):
        new_node = NoLista(v)
        if self.is_empty():
            self.first = new_node
        else:
            last_node = self.last()
            last_node.set_next(new_node)

    def is_equal(self, l):
        current1 = self.first
        current2 = l.first
        while current1 is not None and current2 is not None:
            if current1.get_info() != current2.get_info():
                return False
            current1 = current1.get_next()
            current2 = current2.get_next()
        return current1 is None and current2 is None

    def print_recursive(self):
        self.print_recursive_aux(self.first)
        print()

    def print_recursive_aux(self, node):
        if node is None:
            return
        print(node.get_info(), end=' ')
        self.print_recursive_aux(node.get_next())

    def remove_recursive(self, v):
        self.first = self.remove_recursive_aux(self.first, v)

    def remove_recursive_aux(self, node, v):
        if node is None:
            return None
        if node.get_info() == v:
            return node.get_next()
        node.set_next(self.remove_recursive_aux(node.get_next(), v))
        return node

def main():
    # Example usage
    list_obj = Lista()
    list_obj.insert(1)
    list_obj.insert(2)
    list_obj.insert(3)
    list_obj.print_list()  # Output: 1 2 3

    list_obj.remove(2)
    list_obj.print_list()  # Output: 1 3

if __name__== "__main__":
    main()

File: Data Structure/E3/E3.cpp
#include <iostream>
using namespace std;


class NoLista {
private:
    int info;
    NoLista* next;

public:
    NoLista(int info) {
        this->info = info;
        this->next = nullptr;
    }

    void set_info(int info) {
        this->info = info;
    }

    int get_info() const {
        return info;
    }

    void set_next(NoLista* next) {
        this->next = next;
    }

    NoLista* get_next() const {
        return next;
    }
};

class Lista {
private:
    NoLista* first;

public:
    Lista() {
        this->first = nullptr;
    }

    void insert(int v) {
        NoLista* new_node = new NoLista(v);
        if (is_empty()) {
            first = new_node;
        }
        else {
            NoLista* current = first;
            while (current->get_next() != nullptr) {
                current = current->get_next();
            }
            current->set_next(new_node);
        }
    }

    void print_list() const {
        NoLista* current = first;
        while (current != nullptr) {
            std::cout << current->get_info() << ' ';
            current = current->get_next();
        }
        std::cout << std::endl;
    }

    bool is_empty() const {
        return first == nullptr;
    }

    NoLista* search(int v) const {
        NoLista* current = first;
        while (current != nullptr) {
            if (current->get_info() == v) {
                return current;
            }
            current = current->get_next();
        }
        return nullptr;
    }

    int length() const {
        int count = 0;
        NoLista* current = first;
        while (current != nullptr) {
            count += 1;
            current = current->get_next();
        }
        return count;
    }

    NoLista* last() const {
        NoLista* current = first;
        if (current == nullptr) {
            return nullptr;
        }
        while (current->get_next() != nullptr) {
            current = current->get_next();
        }
        return current;
    }

    void remove(int v) {
        if (is_empty()) {
            return;
        }
        if (first->get_info() == v) {
            NoLista* temp = first;
            first = first->get_next();
            delete temp;
            return;
        }
        NoLista* previous = first;
        NoLista* current = first->get_next();
        while (current != nullptr) {
            if (current->get_info() == v) {
                previous->set_next(current->get_next());
                delete current;
                return;
            }
            previous = current;
            current = current->get_next();
        }
    }

    void clear() {
        NoLista* current = first;
        while (current != nullptr) {
            NoLista* temp = current;
            current = current->get_next();
            delete temp;
        }
        first = nullptr;
    }

    void insert_end(int v) {
        NoLista* new_node = new NoLista(v);
        if (is_empty()) {
            first = new_node;
        }
        else {
            NoLista* last_node = last();
            last_node->set_next(new_node);
        }
    }

    bool is_equal(const Lista& l) const {
        NoLista* current1 = first;
        NoLista* current2 = l.first;
        while (current1 != nullptr && current2 != nullptr) {
            if (current1->get_info() != current2->get_info()) {
                return false;
            }
            current1 = current1->get_next();
            current2 = current2->get_next();
        }
        return current1 == nullptr && current2 == nullptr;
    }

    void print_recursive() const {
        print_recursive_aux(first);
        std::cout << std::endl;
    }

    void print_recursive_aux(NoLista* node) const {
        if (node == nullptr) {
            return;
        }
        std::cout << node->get_info() << ' ';
        print_recursive_aux(node->get_next());
    }

    void remove_recursive(int v) {
        first = remove_recursive_aux(first, v);
    }

    NoLista* remove_recursive_aux(NoLista* node, int v) {
        if (node == nullptr) {
            return nullptr;
        }
        if (node->get_info() == v) {
            NoLista* next_node = node->get_next();
            delete node;
            return next_node;
        }
        node->set_next(remove_recursive_aux(node->get_next(), v));
        return node;
    }
};

int main() {
    // Example usage
    Lista list_obj;
    list_obj.insert(1);
    list_obj.insert(2);
    list_obj.insert(3);
    list_obj.print_list();  // Output: 1 2 3

    list_obj.remove(2);
    list_obj.print_list();  // Output: 1 3

    return 0;
}
